src/storage/DataAccessorBasedStore.ts

```diff
@@ -503,6 +503,11 @@ export class DataAccessorBasedStore implements ResourceStore {
     // Get all values needed for naming the resource
     const isContainer = this.isNewContainer(metadata);
     const slug = metadata.get(SOLID_HTTP.slug)?.value;
+    // Error if slug ends on slash, but no ContainerType Link header is present
+    if (slug && isContainerPath(slug) && !this.hasContainerType(metadata.getAll(RDF.type))) {
+      throw new BadRequestHttpError(`Slugs cannot end with '/' without an HTTP Link header with rel="type" targeting ` +
+      `a valid LDP container type.`);
+    }
     metadata.removeAll(SOLID_HTTP.slug);
 
     let newID: ResourceIdentifier = this.createURI(container, isContainer, slug);
```

Comment:
The problem with this solution is that the `isNewContainer` is going to return an incorrect result since that still checks the slug. That function should be updated to only check the link type. Additionally, the `createUri` function calls `cleanSlug` which also throws an error if there is a slash so now there's some duplication there.

My suggestion would be to instead fix the `isNewContainer` function, merge `createUri` with `cleanSlug`, and update the error there so it only gets thrown in case the result is not a container.

Reply:
Ok, but `cleanSlug()` actually does more: it checks that the slug path does not contain _any_ slashes. So in case of a directory (read: Link header is ok) a slug like `my/test/slug` should still error with the specific `Slugs should not contain slashes` error.

While `myTestSlug/` without a correct Link header could error with `Slugs cannot end with '/' without an HTTP Link header with rel="type" targeting a valid LDP container type.`

So I would:
 * Fix `isNewContainer()` like you said: only check on the link header being present and correct
 * keep `cleanSlug()` as is
 * Revert `createSafeUri()`
 * Modify `createURI()`: test for both cases there:
     1. isContainer == true AND slug.endswitch('/')  => throw my error
     2. cleanSlug() call: throws if slashes are still inside slug with existing error there

Reply:
It looks like this did something similar and suggests slightly different behaviour:
https://github.com/solid/community-server/blob/44d43456b168fc5e692e7ae1fcbe85d4beeb61b0/src/storage/DataAccessorBasedStore.ts#L199-L204

I chose to go for that behaviour, since it is consistent: resource slugs/path don't end in slash, directory slugs/path end in slash with the added requirement of the link header.

Since checking fot this behaviour is required in both `setRepresentation()` and `createURI()`, I've encapsulated it in a new method `validateSlug(isContainer:boolean, suffix?: string): void`.

I am effectively reusing the `BadRequestHttpError` mentioned above.

All unit tests now succeed, but now some integration tests seem to fail, I can't really figure out why though.

Reply:
I am trying to fix most of the obvious failing ones (by adding a Link header, since there are some that just require that)